gnc/perception/sub8_pcl/include/sub8_pcl/geometry.hpp

```diff
@@ -0,0 +1,133 @@
+#pragma once
+#include "typedefs.hpp"
+#include <opencv2/opencv.hpp>
+#include <image_geometry/pinhole_camera_model.h>
+#include <iostream>
+#include <limits>
+
+namespace sub {
+
+// TODO: "further" -- point in xyz, uv for camera_n --> a point $dist further along that ray
+
+// OMP: Compute the point in $cloud that is closest to the ray defined by $direction and $line_pt
+// (For a pc with n cores, this has ~1/nth the runtime of the non-OMP version)
+// (Uses OpenMP to compute in parallel)
+// Compute the minimum over the set of points $cloud: dist(line_pt + direction*t, point)
+// TODO:
+//  ? Additional (wasteful) branching to check if something is actually in front of the ray (Always
+//  true for an image)
+//  - Does squaredNorm improve speed? (This will be miniscule at best, but saves computation)
+//
+// @param[in] cloud The point cloud over which to determine the closest point
+// @param[in] direction The direction of the ray with which we are concerned
+// @param[in] line_pt A point on the line, ideally the origin of the ray
+// @return The index of the point in the target point cloud that is closest to the ray of interest
+template <typename PointT>
+size_t closest_point_index_rayOMP(const pcl::PointCloud<PointT>& cloud,
+                                  const Eigen::Vector3f& direction_pre,
+                                  const Eigen::Vector3f& line_pt) {
+  Eigen::Vector3f direction = direction_pre / direction_pre.norm();
+  double min_distance = std::numeric_limits<double>::infinity();
+  size_t min_index = 0;
+  PointT point;
+  std::vector<double> distances(cloud.points.size(), 10);  // Initialize to value 10
+// Generate a vector of distances
+#pragma omp parallel for
+  for (size_t index = 0; index < cloud.points.size(); index++) {
+    point = cloud.points[index];
+
+    Eigen::Vector3f cloud_pt(point.x, point.y, point.z);
+    Eigen::Vector3f difference = (line_pt - cloud_pt);
+    double distance = (difference - (difference.dot(direction) * direction)).norm();
```

Comment:
Looks like they're equivalent: [implemented, on wikipedia](https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line#Vector_formulation), and [wolfram's formula](https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line#Another_vector_formulation), on wikipedia.